### PR TITLE
HOCS-3316 Only flushing old static lists if retrieval of new ones is …

### PR DESCRIPTION
**Pull request by gardnersam**
…successful

This fixes a bug where occasionally team names would not appear on the dashboard.

Team names are stored in static lists which are refreshed every 15 minutes. This bug occurred when the refresh failed due to the info or casework services not being available as the old behaviour was to delete the static lists, request new ones and repopulate when they arrived, so failure to retrieve new ones left us with nothing. The new behaviour waits to see if new lists are successfully retrieved before deleting and replacing the old data. In the event of failure the preexisting static lists are maintained and the user won't know anything is amiss.

### CODE REVIEW

**Comment by LiamMacP** [comment]
How does this work if user 1 is loading a dashboard and finds no team and while it's flushing the cache user 2 loads the dashboard. Will we have infinite loop of flushing and reloading? We had this issue before which is why we went down the cronjob route.

I don't know if a better route would be to treat this more like a map, we replace if its there, and then flush the values that haven't been found in the static list response?

**Comment by gardnersam** [comment]
@LiamMacP The flush/reload is still only triggered by the cronjob. Users loading the dashboard has no effect on the process. The change here is when the cronjob triggers the refresh the flush is only done if new data has been retrieved and the update is then handled in the same way as before.

**Comment by LiamMacP** [comment]
> @LiamMacP The flush/reload is still only triggered by the cronjob. Users loading the dashboard has no effect on the process. The change here is when the cronjob triggers the refresh the flush is only done if new data has been retrieved and the update is then handled in the same way as before.

Ah my bad, yeah makes sense! Ignore me, nothing to see here...

File: server/services/list/service.js [hunk 1]
```diff
@@ -173,24 +173,23 @@ const getInstance = (requestId, user) => {
...
 const cacheStaticList = async (listId) => {
     const { endpoint, client, adapter } = listRepository.fetch(listId);
     const clientInstance = clientRepository.fetch(client);
     const logger = getLogger();
-    let response;
     try {
-        response = await clientInstance.get(endpoint);
+        const response = await clientInstance.get(endpoint);
+        listCache.flush(listId);
+        const listData = await applyAdapter(response.data, adapter, { logger });
```
Comment (LiamMacP) [inline server/services/list/service.js:187]: Worth moving this up a line, I doubt the applying of the adapter will take long, but would remove the ability to flush and then the adapter to fail...

File: server/services/list/__tests__/service.spec.js [hunk 2]
```diff
@@ -502,16 +502,53 @@ describe('List Service', () => {
...
+            const mockRepositoryFetch = jest.fn();
+            const mockGet = jest.fn();
+            mockGet.mockReturnValue(JSON.parse('{"client": "newclient"}'));
+
+            const mockClientInstance = jest.fn();
+            mockClientInstance.get = mockGet;
+            mockRepositoryFetch.mockReturnValue(mockClientInstance);
 
-            await listService.initialise({}, {}, { listRepository: { flush: mockRepositoryFlush } });
+            await listService.initialise({}, {}, { listRepository: { flush: mockRepositoryFlush, fetch: mockRepositoryFetch, store: mockRepositoryStore } });
```
Comment (zuzak) [inline server/services/list/__tests__/service.spec.js:544]: Split this over several lines.
```suggestion
            await listService.initialise({}, {}, { listRepository: {
                fetch: mockRepositoryFetch,
                flush: mockRepositoryFlush,
                store: mockRepositoryStore
            } });
```

Comment (gardnersam) [reply server/services/list/__tests__/service.spec.js]: Done for this and the other similar line

File: server/services/list/service.js (hunk 1 above)
Comment (gardnersam) [reply server/services/list/service.js:186]: Good shout, I've made the change

**Review by fewtrees-ho — APPROVED** [review]
Approved as per teams call

**Pull request merged**